website/docs/setup/guides/scripting.md

```diff
@@ -0,0 +1,586 @@
+---
+title: Scripting Guide
+sidebar_label: Scripting
+description: Build custom transforms with Vector's scripting capabilities
+status: beta
+---
+
+import Tabs from '@theme/Tabs';
+import TabItem from '@theme/TabItem';
+import Alert from '@site/src/components/Alert';
+import CodeHeader from '@site/src/components/CodeHeader';
+
+When it happens that the built-in [transforms][docs.reference.transforms] are not expressive enough for you needs,
+it is possible to define your own transformation logic using Vector's scripted transforms.
+
+Currently there are two supported scripted transforms, [`lua`][docs.reference.transforms.lua] and
+[`javascript`][docs.reference.transforms.lua]. For most use cases they can be used interchangeably, so one
+can chose between them based on personal preferences.
+
+## Motivating Example
+
+Let us take a look at a simple transform which enumerates incoming log events and then sends events with even
+numbers to one lane and
+events with odd numbers to another.
+
+<Tabs
+  block={true}
+  defaultValue="lua"
+  urlKey="lang"
+  values={[
+    { label: 'Lua', value: 'lua'},
+    { label: 'JavaScript', value: 'javascript'},
+  ]}>
+
+<TabItem value="lua">
+
+<CodeHeader fileName="splitter.lua" />
+
+```lua
+counter = 0
+
+function handler (event)
+  if not event.log then
+    error("only log events are supported")
+  end
+
+  counter = counter + 1
+  event.log.id = counter
+
+  if counter % 2 == 0 then
+    event.lane = "even"
```

Comment:
I didn't see anywhere in this guide where the `lane` field is actually used. It would be more clear to follow through with this example.

I'm also curious if you saw #1942? It might be easier to provide a way for scripting transforms to define outputs versus pairing it with the `swimlandes` transform.

Reply:
I saw #1942 and I want to solve it by allowing the user to optionally add `lane` field to the top level of the event. If the user adds this field, the effect should be the same.

Reply:
I see. I'd prefer that we prefix special field with `_`, `@`, or some other character. To throw another edge case at you, what if a user wanted to emit an event to 2 lanes?

Reply:
> I'd prefer that we prefix special field with _, @, or some other character.

It is already a kind of "meta" field, as it is not visible to the downstream components which see only the event which is inside the value of the `log` key. For example, the original produced by say `stdin` source looks like this:

```json
{
  "message": "message",
  "timestamp": "2020-03-02T15:45:56.751Z",
  "host": "localhost"
}
```

Then it passed to the `lua` transform as a table

```lua
{
  log = {
    message = "message",
    timestamp = "2020-03-02T15:45:56.751Z",
    host = "localhost"
  }
}
```

Setting the lane by running `event.lane = "lane_name"` results in this:

```lua
{
  log = {
    message = "message",
    timestamp = "2020-03-02T15:45:56.751Z",
    host = "localhost"
  },
  lane = "lane_name"
}
```

After this table is returned from the transform, the `log` part is extracted to be sent to the downstream components reading from `transform_name.lane_name`. So if there is a downstream `console` sink reading from `transform_name.lane_name` it would receive

```json
{
  "message": "message",
  "timestamp": "2020-03-02T15:45:56.751Z",
  "host": "localhost"
}
```

Furthermore, I'm thinking about another "meta" field visible only to the transform, `input`, which would contain the name of the component which produced the event. But it would require changes on other parts of code to implement.

Reply:
That makes more sense, I forgot the type data was nested. I still find the `lanes` approach to be somewhat awkward. IMO, it feels more natural to pair this with the `return` statement or an explicit function. For example:

```lua
return {event, "lane1", "lane2"}
```

And returning just `event` would work as well:

```lua
return event
```

Is your intent to avoid having to implement a common API across all scripting transforms? I assume adding a `lane` field would make it easier to share this logic across different scripting transforms?

Reply:
I'm experimenting with the approach proposed in #1942, with the transform pushing the events using an `emit` function. But then I think it makes sense to make the handlers closures taking which take entire input stream and then iterate over it and emit output events.

For example, if the handler takes `input` parameter which is an [iterator](https://www.lua.org/pil/7.1.html) and the `emit` function, then the entire handler can look like this:

```lua
function (input, emit)
  -- init
  counter = 0

  -- process
  for event in input do
    counter = counter + 1
    event.log.id = counter

    if counter % 2 == 0 then
      emit(event, 'even')
    else
      emit(event, 'odd')
    end
  end

  -- clean up
  emit {
    log = {
      message = "good bye",
    }
  }
```

Which implements the [true iterator](https://www.lua.org/pil/7.5.html) pattern from Lua. However, my concern with this approach is that it is not possible to produce an event if the waiting takes too long. So we need a more flexible approach, which is also not overcomplicated in the simple cases, I consider a few of them.

In general, I think if add such large changes, it is worth to create an RFC on the new programming model which would solve all our problems.

Reply:
Agree that we’d need a RFC for that. I prefer your original proposal over this direction, as I find it to be less straightforward. Happy to have others weigh in if you disagree. And appreciate you trying that out.